### ISSUE TITLE
`jnp.log2` does not compute integer powers of 2 exactly

### ISSUE THREAD

**Issue by gonnet** [jax-ml/jax#40419]
### Description

The `jnp.log2` function fails to return exact values for several exact powers of two for `jnp.float32` and `jnp.bfloat16`.

The following code illustrates this:
```py
for dtype in {jnp.float32, jnp.bfloat16}:
  dinfo = jnp.finfo(dtype)
  e =  jnp.arange(dinfo.minexp, dinfo.maxexp)
  values = jnp.ldexp(1.0, e).astype(dtype)
  err_inds = jnp.where(jnp.log2(values) != e)[0]
  print(f'{dtype.__name__}: jnp.log2(2^k) == k failed for {err_inds.size} of {e.size} values of k.')
```
which produces the following result:
```
bfloat16: jnp.log2(2^k) == k failed for 105 of 254 values of k.
float32: jnp.log2(2^k) == k failed for 46 of 254 values of k.
```

It looks like `jnp.log2(x)` is implemented as `jnp.log(x) / jnp.log(2)`, which fails to recover the exact solution since neither `log(2^k)` nor `log(2)` are represented exactly for any `k != 0`.

I'm guessing the same goes for `jnp.log10`.

One way to avoid this problem (for `jnp.log2` at least) would be to do the following:
```py
    def log2(x: jax.Array) -> jax.Array:
      """Computes the base-2 logarithm of `x`."""
      m, e = jnp.frexp(x)
      return e + jnp.log(m) * (1.0 / math.log(2.0))
```
which, inserted into the above code, produces the output:
```
bfloat16: log2(2^k) == k failed for 0 of 254 values of k.
float32: log2(2^k) == k failed for 0 of 254 values of k.
```
but I'm not accuracy/performance tradeoff is worth it for JAX users.

### System info (python version, jaxlib version, accelerator, etc.)

```
jax:    0.11.2
jaxlib: 0.11.2
numpy:  2.5.2
python: 3.14.5 (stable, redacted, redacted) [Clang 9999.0.0 (75a45c373407c13a44c7abb28a78d891a97fe665)]
device info: cpu-1, 1 local devices"
process_count: 1
platform: uname_result(system='Linux', node='ti-vm-4555d52793ffea74-1abc8647003', release='6.6.106-smp-1302.23.413.7', version='#1 [v6.6.106-1302.23.413.7] SMP @1776377115', machine='x86_64')
```

**Comment by jakevdp** [comment]
Hi - thanks for the report!

In some senses, this is behaving as expected: `log2` is a compound floating point operation, and floating point operations necessarily have some rounding error.

On the other hand, as you point out, we could change the implementation of `log2` to make it slightly more accurate. However, the change you suggest comes at a cost; compare the jaxprs of the two approaches:
```python
In [1]: import jax

In [2]: import jax.numpy as jnp

In [3]: import math

In [4]: def log2(x: jax.Array) -> jax.Array:
   ...:   """Computes the base-2 logarithm of `x`."""
   ...:   m, e = jnp.frexp(x)
   ...:   return e + jnp.log(m) * (1.0 / math.log(2.0))
   ...: 

In [5]: jax.make_jaxpr(jnp.log2)(2.0)
Out[5]: 
{ lambda ; a:f32[]. let
    b:f32[] = log a
    c:f32[] = log 2.0:f32[]
    d:f32[] = div b c
  in (d,) }

In [6]: jax.make_jaxpr(log2)(2.0)
Out[6]: 
let jaxpr = { lambda ; a:f32[]. let
    b:f32[] = abs a
    c:f32[] = convert_element_type[new_dtype=float32 weak_type=False] b
    d:bool[] = lt c 1.1754943508222875e-38:f32[]
    e:f32[] = mul a 8388608.0:f32[]
    f:f32[] = jit[
      name=_where
      jaxpr={ lambda ; d:bool[] e:f32[] a:f32[]. let
          f:f32[] = select_n d a e
        in (f,) }
    ] d e a
    g:i32[] = jit[name=_where jaxpr=_where] d -23:i32[] 0:i32[]
    h:i32[] = bitcast_convert_type[new_dtype=int32] f
    i:i32[] = shift_right_arithmetic h 23:i32[]
    j:i32[] = and i 255:i32[]
    k:i32[] = sub j 127:i32[]
    l:i32[] = add k 1:i32[]
    m:i32[] = add g l
    n:i32[] = and h -2139095041:i32[]
    o:i32[] = or n 1056964608:i32[]
    p:f32[] = bitcast_convert_type[new_dtype=float32] o
    q:bool[] = jit[
      name=isinf
      jaxpr={ lambda ; a:f32[]. let
          r:f32[] = abs a
          q:bool[] = eq r inf:f32[]
        in (q,) }
    ] a
    s:bool[] = ne a a
    t:bool[] = convert_element_type[new_dtype=bool weak_type=False] s
    u:bool[] = or q t
    v:bool[] = eq a 0.0:f32[]
    w:bool[] = convert_element_type[new_dtype=bool weak_type=False] v
    x:bool[] = or u w
    y:i32[] = jit[name=_where jaxpr=_where] x 0:i32[] m
    z:f32[] = jit[
      name=_where
      jaxpr={ lambda ; x:bool[] a:f32[] p:f32[]. let
          ba:f32[] = convert_element_type[new_dtype=float32 weak_type=False] a
          z:f32[] = select_n x p ba
        in (z,) }
    ] x a p
  in (z, y) } in
let _where = { lambda ; bb:bool[] bc:i32[] bd:i32[]. let
    be:i32[] = select_n bb bd bc
  in (be,) } in
{ lambda ; bf:f32[]. let
    bg:f32[] bh:i32[] = jit[
      name=frexp
      jaxpr={ lambda ; bf:f32[]. let
          bg:f32[] bh:i32[] = custom_jvp_call[
            name=_frexp
            call_jaxpr=jaxpr
            jvp=_frexp_jvp
            symbolic_zeros=False
          ] bf
        in (bg, bh) }
    ] bf
    bi:f32[] = log bg
    bj:f32[] = mul bi 1.4426950216293335:f32[]
    bk:f32[] = convert_element_type[new_dtype=float32 weak_type=False] bh
    bl:f32[] = add bk bj
  in (bl,) }
```
I think we can agree that the proposal adds some complexity that some users may not appreciate if they're not looking at bitwise equality of specific special values!

Given this, I lean toward keeping the status quo – an efficient implementation that is correct to within typical rounding errors that users of floating point operations should expect.

**Comment by jakevdp** [comment]
One small change we might make: if we change this line: https://github.com/jax-ml/jax/blob/406ff19433f009ad40ba126e21d2f966b71f79c3/jax/_src/numpy/ufuncs.py#L2820
to this:
```python
out = lax.mul(lax.log(x), _constant_like(x, 1. / np.log(2)))
```
then the bfloat16 mismatches go from 105 down to 72 by virtue of computing the `1/log(2)` factor on host at 64-bit precision, and we don't have any tradeoffs in terms of runtime complexity.

**Comment by gonnet** [comment]
> One small change we might make: 

Yeah, there's absolutely no reason for computing `jnp.log(2)` on the fly. Also, division is significantly more costly than multiplication, so it's a solid win.

You could arguably also split `scale = 1.0 / math.log(2)` into two constants, e.g. `scale = scale_lo + scale_hi` and shave off an extra bit (literally) of error by doing
```py
  scale = 1.0 / math.log(2)
  scale_hi = _constant_like(x, scale)
  scale_lo = _constant_like(x, scale - scale_hi)  # Not even sure we can do this?
  log_x = lax.log(x)
  out = lax.add(lax.mul(log_x, scale_lo), lax.mul(log_x, scale_hi))
```
WDYT?

**Comment by gonnet** [comment]
> I think we can agree that the proposal adds some complexity that some users may not appreciate if they're not looking at bitwise equality of specific special values!

Ugh, yes. I was assuming that `frexp` was a low-level function with complexity comparable to `log`, but it looks like JAX does it "by hand", i.e. this is not delegated to an XLA op? IIUC `immintrin.h` defines the `getexp`/`getmant` intrinsics, along with the`scalef` intrinsics that implement `ldexp`, i.e. the inverse of `frexp`, so technically this might be doable more efficiently?

What bothers me the most, though, is that there is no "native" `log2` in JAX and we even have to do these tricks to begin with. If you look at XNNPACK's [`f32-vlog`](https://github.com/google/XNNPACK/blob/master/src/f32-vlog/rational-3-3.c.in) and (especially) [`f32-vexp`](https://github.com/google/XNNPACK/blob/master/src/f32-vexp/rational-3-2.c.in), the "natural" logarithm/exponential is computed via the base-2 logarithm/exponential because we get most of it for free since we're in a base-2 representation to begin with.

All this to say that a "native" `log2` kernel would be not only faster, but also more accurate than the corresponding "log" kernel. But can JAX rely on all its backends exposing something like this?

The reason XNNPACK doesn't support `log2` is because nobody ever asked for it, but maybe nobody asked because they didn't know it would be easier than "natural" `log`?

**Comment by hawkinsp** [comment]
I think we should (when we have time) make log2 primitive in JAX, StableHLO and HLO. That is because `log2` is actually the native log op on a bunch of hardware platforms.

And if we're doing that, we should do exp2 and log10 as well.

**Comment by gonnet** [comment]
I'm looking at the [documentation for `_mm_getmant_ps`](https://www.intel.com/content/www/us/en/docs/intrinsics-guide/index.html#ig_expand=3603&text=_mm_get*_ps&cats=Miscellaneous)  for the first time and, OMG, whoever implemented this fully deserves all the (numerical) love and respect in the world.

This thing covers **every**. **single**. (weird/wonderful) **usecase**. **imaginable**. **ever**. 😲

**Comment by gonnet** [comment]
(I mean, being able to select the interval is awesome, but being able to select the **sign behaviour** is just 🧑‍🍳💋)

**Comment by jakevdp** [comment]
> Ugh, yes. I was assuming that `frexp` was a low-level function with complexity comparable to `log`, but it looks like JAX does it "by hand", i.e. this is not delegated to an XLA op?

Right – there is no XLA op for `frexp`, and the only reason we define `jax.numpy.frexp` is for parity with the NumPy API (that was a goal early in the project; if we were doing it over I'd probably opt to skip APIs like this where the implementation available in XLA doesn't actually make sense!)

**Comment by gonnet** [comment]
@jakevdp Would it make sense to ask XLA to add ops for `frexp` and `ldexp`?

**Comment by hawkinsp** [comment]
Sent https://github.com/openxla/stablehlo/pull/3007 as a first step towards adding a native log2/exp2.
